### DISCUSSION TITLE
Can I debug the code using printk() in this open source code like in normal linux kernel code？

### DISCUSSION THREAD

**Discussion by wenhaopang** [NVIDIA/open-gpu-kernel-modules#347] [General]
Can I debug the code using _printk()_ in this open source code like in normal linux kernel code？

**Comment by mtijanic** [comment]
Of course! We don't include any linux specific headers in the code that is meant to be cross-platform, but for local debugging that's no issue.

If you're in the `nvidia.ko` portion of the codebase (`src/nvidia/...`), `portDbgPrintf()` is generally a thin wrapper around `printk()` that you can readily use without including anything. Also `NV_PRINTF(LEVEL_ERRORS, "...")` behaves very similarly to printk() with some added functionality.

**Reply by wenhaopang** [reply]
Thanks for your guidance.

**Reply by wenhaopang** [reply]
Can I also view the output of `portDbgPrintf()` and `NV_PRINTF()` via `dmesg` or `cat /proc/kmsg`.

**Reply by mtijanic** [reply]
Yes. `portDbgPrintf()` should always show up in `dmesg`,  while `NV_PRINTF()` by default only prints on `LEVEL_ERROR` and above. You can control how/when `NV_PRINTF()` actually prints using RmMsg, as described here: https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/197

**Reply by wenhaopang** [reply]
I understand, thanks.
